syntax = "proto3";

import "gogo.proto";

package trade.service.v1;

option go_package = "v1";

// Trade grpc
service StreamEngineService {
  // 获取单次行情
  rpc GetQuote(GetQuoteReq) returns (QuoteData);
  // 订阅聚合行情
  //rpc SubscribeQuote(SubscribeQuoteReq) returns (stream QuoteData);
  rpc MultiSubscribeQuote(SubscribeQuoteReq) returns (stream MultiQuoteData);
}

message SubscribeQuoteReq {

}

message GetQuoteReq {
  string exchange = 1 [(gogoproto.customname) = "Exchange", (gogoproto.jsontag) = "exchange"];
  string symbol = 2 [(gogoproto.customname) = "Symbol", (gogoproto.jsontag) = "symbol"];
}

message MultiQuoteData {
  repeated QuoteData quotes = 1 [(gogoproto.customname) = "Quotes", (gogoproto.jsontag) = "quotes"];
}

message QuoteData {
  string symbol = 1 [(gogoproto.customname) = "Symbol", (gogoproto.jsontag) = "symbol"];
  int64 msg_seq = 2 [(gogoproto.customname) = "MsgSeq", (gogoproto.jsontag) = "msg_seq"];
  string time = 3 [(gogoproto.customname) = "Time", (gogoproto.jsontag) = "time", (gogoproto.casttype) = "wx/pkg/time.Time"];
  string time_arrive = 4 [(gogoproto.customname) = "TimeArrive", (gogoproto.jsontag) = "time_arrive", (gogoproto.casttype) = "wx/pkg/time.Time"];
  repeated DepthLevel ask_depth = 5 [(gogoproto.customname) = "AskDepth", (gogoproto.jsontag) = "ask_depth"];
  repeated DepthLevel bid_depth = 6 [(gogoproto.customname) = "BidDepth", (gogoproto.jsontag) = "bid_depth"];
  bool is_snap = 7 [(gogoproto.customname) = "IsSnap", (gogoproto.jsontag) = "is_snap"];
}

message DepthLevel {
  Decimal price = 1 [(gogoproto.customname) = "Price", (gogoproto.jsontag) = "price"];
  repeated DepthVolume data = 2 [(gogoproto.customname) = "Data", (gogoproto.jsontag) = "data"];
}

message Decimal {
  int64 value = 1 [(gogoproto.customname) = "Value", (gogoproto.jsontag) = "value"];
  int32 base = 2 [(gogoproto.customname) = "Base", (gogoproto.jsontag) = "base"];
}

message DepthVolume {
  double volume = 1 [(gogoproto.customname) = "Volume", (gogoproto.jsontag) = "volume"];
  string exchange = 2 [(gogoproto.customname) = "Exchange", (gogoproto.jsontag) = "exchange"];
}